std/zlib.d

```diff
@@ -604,6 +605,8 @@ class UnCompress
 
         if (zs.avail_in)
             buf = zs.next_in[0 .. zs.avail_in] ~ cast(ubyte[]) buf;
+        else //avoid having buf around, as it causes problems with reusing buffers
+            buf = buf.idup();
```

Comment:
idup isn't free, why we can't reuse buffer?

Reply:
zs.next_in is still a pointer to the same buffer as before
if after the first call not all of the buffer has been consumed (zs.avail_in > 0) and the same buffer has been used to read data (eg. with std.stdio.File.rawRead()), buf will now contain overridden data where zlib expects to find the old non-consumed data

Example:
```
auto uc = new UnCompress();
auto file = File("test.dat", "rb");

ubyte[512] buffer;
ubyte[] usedBuffer;

while(true) {
    usedBuffer = file.rawRead(buffer);
    if(usedBuffer.length == 0)
        break;
    uc.uncompress(usedBuffer);
}
uc.flush();
```

Reply:
I don't think this is the right fix. This will introduce a performance regression into programs that previously worked correctly with no simple workaround.

Perhaps there would be a `compress` / `uncompress` overload that takes an immutable buffer, which would carry with itself the guarantee of arbitrary lifetime.

However, looking at the code, I'm skeptical this function is still correct as it is now. If the output buffer is consumed entirely before the input buffer is, there don't seem to be any provisions to either extend the output buffer and retry, or notify the caller that not the entire input buffer was consumed.

Reply:
I currently have the following additional suggestions:

- a new overload of `uncompress()` that takes an additional `out c_ulong numConsumed` and/or
- deferring the call to `idup()` until after calling `inflate()`

The former is my favorite as it preserves current behaivour and API, while imposing no big performance hit through unnecessary allocation

Reply:
Can `uncompress` not loop while the input buffer is not fully consumed, growing the output buffer (`destbuf`) as needed? Or will zlib refuse to consume all input (even with a sufficient output buffer) in some situations?

Reply:
According to the [zlib manual](https://zlib.net/manual.html#Basic)

> The detailed semantics are as follows. inflate performs one or both of the following actions:
> - Decompress more input starting at next_in and update next_in and avail_in accordingly. If not all input can be processed (because there is not enough room in the output buffer), then next_in and avail_in are updated accordingly, and processing will resume at this point for the next call of inflate().
> - Generate more output starting at next_out and update next_out and avail_out accordingly. inflate() provides as much output as possible, until there is no more input data or no more space in the output buffer (see below about the flush parameter).

Expanding the buffer should lead to consuming all of the input buffer, although it will require (few) reallocations.

Should I implement this (I'll open a new PR if so)?

Reply:
> Expanding the buffer should lead to consuming all of the input buffer, although it will require (few) reallocations.

Reallocations are unavoidable without decoding the input twice. Perhaps `appender` would be suitable here.

Some file formats encode the uncompressed length (gzip), but this does not apply to an arbitrary stream of compressed data, of course.

> Should I implement this (I'll open a new PR if so)?

I think it would be more correct than the solution here.

A test would be good too - as I understand, the current implementation misbehaves when the decompressed data is over twice longer than the compressed data, so a test that would fail in the current implementation would confirm it.